test/common/upstream/health_check_fuzz.cc

```diff
@@ -0,0 +1,120 @@
+#include "test/fuzz/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+HealthCheckFuzz::HealthCheckFuzz() {}
+
+void HealthCheckFuzz::allocHealthCheckerFromProto(
+    const envoy::config::core::v3::HealthCheck& config) {
+  health_checker_ = std::make_shared<TestHttpHealthCheckerImpl>(
+      *cluster_, config, dispatcher_, runtime_, random_,
+      HealthCheckEventLoggerPtr(event_logger_storage_.release()));
+  ENVOY_LOG_MISC(trace, "Created Test Health Checker");
+}
+
+void HealthCheckFuzz::initialize(test::common::upstream::HealthCheckTestCase input) {
+  allocHealthCheckerFromProto(input.health_check_config());
+  addCompletionCallback();
+  if (input.second_host()) {
+    cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+        Upstream::makeTestHost(cluster_->info_, "tcp://127.0.0.1:80"),
+        Upstream::makeTestHost(cluster_->info_, "tcp://127.0.0.1:81")};
+  } else {
+    cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+        Upstream::makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  }
+  expectSessionCreate();
+  expectStreamCreate(0);
+  if (input.second_host()) {
+    ENVOY_LOG_MISC(trace, "Created second host.");
+    second_host_ = true;
+    expectSessionCreate();
+    expectStreamCreate(1);
+  }
+  health_checker_->start();
+  replay(input);
+}
+
+void HealthCheckFuzz::respondHeaders(
+    test::fuzz::Headers headers, absl::string_view status,
+    bool respond_on_second_host) {
+  std::unique_ptr<Http::TestResponseHeaderMapImpl> response_headers =
+      std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(headers, {}, {}));
+  
+  // If Fuzzer created a status header - replace it with validated status
+  response_headers->setStatus(status);
+
+  int index = (second_host_ && respond_on_second_host) ? 1 : 0;
+
+  test_sessions_[index]->stream_response_callbacks_->decodeHeaders(
+        std::move(response_headers), true);
+
+
+  //Asserts correct response based on logic in the codebase
+  uint64_t response_code = std::stoi(static_cast<std::string>(status));
+  if (!health_checker_->httpStatusChecker().inRange(response_code)) {
+      ENVOY_LOG_MISC(trace, "Response code not in healthy range. Host is unhealthy");
+      ASSERT(cluster_->prioritySet().getMockHostSet(0)->hosts_[index]->health() == Host::Health::Unhealthy);
+  }
+
+  if (response_headers->EnvoyDegraded() != nullptr) {
+    ENVOY_LOG_MISC(trace, "Replied that the host is degraded");
+    ASSERT(cluster_->prioritySet().getMockHostSet(0)->hosts_[index]->health() == Host::Health::Degraded);
+  }
+
+  //No clauses that represent the host not being healthy
+  ENVOY_LOG_MISC(trace, "Replied that the host is Healthy");
+  ASSERT(cluster_->prioritySet().getMockHostSet(0)->hosts_[index]->health() == Host::Health::Healthy);
+}
+
+void HealthCheckFuzz::streamCreate(bool create_stream_on_second_host) {
+  ENVOY_LOG_MISC(trace, "Created a new stream on host 1.");
```

Comment:
Should probably be either 0 or 1 depending on second_host and create_stream_on_second_host.